### PR TITLE
[WIP] The Road to Compiling (Sponge branch)

### PR DESCRIPTION
**Pull request by hsyyid**
This PR will (eventually) make WorldGuard for Sponge compile.

Any input is welcome. :smile:

### CODE REVIEW

File: src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java [hunk 1]
```diff
@@ -19,9 +19,21 @@
 
 package com.sk89q.worldguard.sponge.listener;
 
+<<<<<<< HEAD
+import static com.sk89q.worldguard.sponge.util.WorldEditTransforms.toVector;
```
Comment (clienthax) [inline src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java:22]: Derp

File: src/main/java/com/sk89q/worldguard/chest/SignChestProtection.java [hunk 2]
```diff
@@ -19,7 +19,10 @@
 
 package com.sk89q.worldguard.chest;
 
+<<<<<<< HEAD
+=======
 import java.util.Optional;
+>>>>>>> upstream/sponge
```
Comment (clienthax) [inline src/main/java/com/sk89q/worldguard/chest/SignChestProtection.java]: derp

File: src/main/java/com/sk89q/worldguard/protection/flags/LazyLocation.java [hunk 3]
```diff
@@ -20,12 +20,17 @@
 package com.sk89q.worldguard.protection.flags;
 
 import com.flowpowered.math.vector.Vector3d;
+<<<<<<< HEAD
+=======
 import java.util.Optional;
+>>>>>>> upstream/sponge
```
Comment (clienthax) [inline src/main/java/com/sk89q/worldguard/protection/flags/LazyLocation.java]: derp

File: src/main/java/com/sk89q/worldguard/protection/flags/IntegerFlag.java [hunk 4]
```diff
@@ -36,7 +36,11 @@ public IntegerFlag(String name) {
     }
 
     @Override
+<<<<<<< HEAD
+    public Integer parseInput(WorldGuardPlugin plugin, CommandSource sender,String input) throws InvalidFlagFormat {
+=======
     public Integer parseInput(WorldGuardPlugin plugin, CommandSource sender, String input) throws InvalidFlagFormat {
```
Comment (clienthax) [inline src/main/java/com/sk89q/worldguard/protection/flags/IntegerFlag.java]: derp

File: src/main/java/com/sk89q/worldguard/protection/flags/EnumFlag.java [hunk 5]
```diff
@@ -19,12 +19,16 @@
 
 package com.sk89q.worldguard.protection.flags;
 
+<<<<<<< HEAD
+=======
 import java.util.Optional;
```
Comment (clienthax) [inline src/main/java/com/sk89q/worldguard/protection/flags/EnumFlag.java]: derp

**Comment by bloodmc** [comment]
@hsyyid squash some of these commits

**Comment by hsyyid** [comment]
@bloodmc Okay, I'll squash em all

File: src/main/java/com/sk89q/worldguard/protection/managers/RegionManager.java [hunk 6]
```diff
@@ -54,396 +54,450 @@
...
-     * Helper method for {@link #getApplicableRegions(Vector3d)} using Bukkit
-     * locations.
-     *
-     * @param loc the location
-     * @return an {@code ApplicableRegionSet}
-     */
-    public ApplicableRegionSet getApplicableRegions(Location loc) {
-        return getApplicableRegions(loc.getPosition());
-    }
+public final class RegionManager
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/protection/managers/RegionManager.java]: Why did you mess with the formatting on this class?

File: src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java [hunk 7]
```diff
@@ -224,7 +235,7 @@ public void onUseItem(UseItemEvent event) {
     }
 
     @Listener
-    public void onPlayerDropItem(DropItemStackEvent.Pre event) {
+    public void onPlayerDropItem(DropItemEvent.Dispense event) {
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java]: This should be DropItemEvent to handle Destruct too

File: src/main/java/com/sk89q/worldguard/sponge/listener/debounce/BlockBlockPair.java [hunk 8]
```diff
@@ -19,14 +19,14 @@
...
-import org.bukkit.block.Block;
+import org.spongepowered.api.block.BlockSnapshot;
 
 public class BlockBlockPair {
 
     private final BlockMaterialPair block1;
     private final BlockMaterialPair block2;
 
-    public BlockBlockPair(Block block1, Block block2) {
+    public BlockBlockPair(BlockSnapshot block1, BlockSnapshot block2) {
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/debounce/BlockBlockPair.java]: This class isn't needed. Use 

```
Transaction<BlockSnapshot>
```

File: src/main/java/com/sk89q/worldguard/sponge/listener/debounce/BlockMaterialPair.java [hunk 9]
```diff
@@ -19,20 +19,20 @@
...
+
 public class BlockMaterialPair {
 
-    private final Block block;
-    private final Material material;
+    private final BlockSnapshot block;
+    private final BlockType material;
 
-    public BlockMaterialPair(Block block) {
+    public BlockMaterialPair(BlockSnapshot block) {
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/debounce/BlockMaterialPair.java]: This class isn't needed, you just need a single BlockSnapshot.

File: pom.xml [hunk 10]
```diff
@@ -447,4 +447,4 @@
       </build>
     </profile>
   </profiles>
-</project>
+</project>
```
Comment (bloodmc) [inline pom.xml:450]: Unneeded formatting change

File: src/main/java/com/sk89q/worldguard/protection/managers/index/HashMapIndex.java [hunk 11]
```diff
@@ -291,4 +277,25 @@ public HashMapIndex get() {
         }
     }
 
+	@Override
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/protection/managers/index/HashMapIndex.java]: Fix formatting

**Comment by hsyyid** [comment]
@bloodmc I have no clue why it's doing that to the POM.XML, must be something dumb about my IDE.

File: src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java [hunk 12]
```diff
@@ -254,70 +261,81 @@ public void onBlockDispense(BlockDispenseEvent event) {
...
-            WorldConfiguration wcfg = cfg.get(entity.getWorld());
-            LocalPlayer localPlayer = getPlugin().wrapPlayer(player);
-
-            if (wcfg.getBlacklist() != null && !wcfg.getBlacklist().check(
-                    new ItemAcquireBlacklistEvent(localPlayer, toVector(entity.getLocation()), createTarget(item)), false, false)) {
-                event.setCancelled(true);
-
-                if (inventory.getHolder().equals(player)) {
-                    event.setCurrentItem(null);
+    public void onInventoryClick(InteractInventoryEvent event) {
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/BlacklistListener.java]: This should be ClickInventoryEvent

File: src/main/java/com/sk89q/worldguard/sponge/listener/debounce/legacy/BlockEntityEventDebounce.java [hunk 13]
```diff
@@ -32,18 +32,18 @@ public BlockEntityEventDebounce(int debounceTime) {
...
-    public <T extends Event & Cancellable> void debounce(Block block, Entity entity, Cancellable originalEvent, T firedEvent) {
+    public <T extends Event & Cancellable> void debounce(BlockSnapshot block, Entity entity, Cancellable originalEvent, T firedEvent) {
         super.debounce(new Key(block, entity), originalEvent, firedEvent);
     }
 
     protected static class Key {
-        private final Block block;
-        private final Material blockMaterial;
+        private final BlockSnapshot block;
+        private final BlockType blockType;
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/debounce/legacy/BlockEntityEventDebounce.java]: This should be removed as it is already part of BlockSnapshot.

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 14]
```diff
@@ -286,9 +297,8 @@ public boolean apply(Location target) {
         });
     }
 
-    @EventHandler(ignoreCancelled = true)
+    @Listener
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java]: This listener should handle multiple entity spawns, not just one.

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 15]
```diff
@@ -333,13 +343,12 @@ public void onSpawnEntity(SpawnEntityEvent event) {
         }
     }
 
-    @EventHandler(ignoreCancelled = true)
+    @Listener
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java:348]: DestroyEntityEvent does not exist, this should be DestructEntityEvent

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 16]
```diff
@@ -190,9 +203,8 @@ public boolean apply(Location target) {
         });
     }
 
-    @EventHandler(ignoreCancelled = true)
+    @Listener
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java:205]: BreakBlockEvent ? This no longer exists and should be ChangeBlockEvent.Break

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 17]
```diff
@@ -138,21 +152,20 @@ private RegionAssociable createRegionAssociable(Cause cause) {
...
             RegionQuery query = getPlugin().getRegionContainer().createQuery();
-            return new DelayedRegionOverlapAssociation(query, ((Block) rootCause).getLocation());
+            return new DelayedRegionOverlapAssociation(query, ((BlockSnapshot) rootCause).getLocation().get());
         } else {
             return Associables.constant(Association.NON_MEMBER);
         }
     }
 
-    @EventHandler(ignoreCancelled = true)
+    @Listener
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java]: PlaceBlockEvent no longer exists, should be ChangeBlockEvent.Place

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 18]
```diff
@@ -229,13 +241,12 @@ public boolean apply(Location target) {
...
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java]: UseBlockEvent no longer exists. Should be InteractBlockEvent.Secondary

File: src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java [hunk 19]
```diff
@@ -369,13 +378,12 @@ public void onDestroyEntity(DestroyEntityEvent event) {
...
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/RegionProtectionListener.java]: UseEntityEvent no longer exists, should be InteractEntityEvent.Secondary

File: src/main/java/com/sk89q/worldguard/sponge/listener/WorldGuardBlockListener.java [hunk 20]
```diff
@@ -83,13 +87,15 @@ public void onBlockBreak(BreakBlockEvent event) {
      * Called when fluids flow.
      */
     @Listener
-    public void onBlockFromTo(final PlaceBlockEvent event) {
-        Location<World> block = event.getCause().getFirst(Location.class).orNull();
-        if (block == null) return;
+    public void onBlockFromTo(final ChangeBlockEvent.Place event) {
+        Location<World> block = event.getCause().first(Location.class).orElse(null);
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/WorldGuardBlockListener.java]: Location will never be in the cause, this should be looping through each

```
Transaction<BlockSnapshot>
```

**Comment by hsyyid** [comment]
@bloodmc All of your comments are now fixed :smile:

**Comment by bloodmc** [comment]
@hsyyid definitely not fixed, you need to fix RegionProtectionListener. Many events in that file do not exist.

**Comment by hsyyid** [comment]
@bloodmc Erm... all I said was I fixed your comments - not I fixed everything. Also you commented on the wrong lines, making it look like it's not fixed.

File: src/main/java/com/sk89q/worldguard/sponge/listener/WorldGuardPlayerListener.java [hunk 21]
```diff
@@ -185,74 +181,82 @@ public void onPlayerLogin(ClientConnectionEvent.Auth event) {
...
-                if (set.size() > 0) {
-                    player.sendMessage(Texts.of(TextColors.YELLOW, "Can you build? "
-                            + (set.testState(localPlayer, DefaultFlag.BUILD) ? "Yes" : "No")));
-
-                    StringBuilder str = new StringBuilder();
-                    for (Iterator<ProtectedRegion> it = set.iterator(); it.hasNext(); ) {
-                        str.append(it.next().getId());
-                        if (it.hasNext()) {
-                            str.append(", ");
+    public void handleBlockRightClick(InteractBlockEvent event) {
```
Comment (bloodmc) [inline src/main/java/com/sk89q/worldguard/sponge/listener/WorldGuardPlayerListener.java]: This should be InteractBlockEvent.Secondary

**Pull request merged**